src/contracts/core/DelegationManager.sol

```diff
@@ -45,107 +45,155 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
         unchecked {
-            nonces[staker] = nonce + 1;
+            stakerNonce[staker] = currentStakerNonce + 1;
         }
 
-        bytes32 digestHash;
-        if (block.chainid != ORIGINAL_CHAIN_ID) {
-            bytes32 domain_separator = keccak256(abi.encode(DOMAIN_TYPEHASH, keccak256(bytes("EigenLayer")), block.chainid, address(this)));
-            digestHash = keccak256(abi.encodePacked("\x19\x01", domain_separator, structHash));
-        } else{
-            digestHash = keccak256(abi.encodePacked("\x19\x01", DOMAIN_SEPARATOR, structHash));
-        }
+        // calculate the digest hash
+        bytes32 stakerDigestHash = keccak256(abi.encodePacked("\x19\x01", domainSeparator(), stakerStructHash));
 
-        /**
-         * check validity of signature:
-         * 1) if `staker` is an EOA, then `signature` must be a valid ECDSA signature from `staker`,
-         * indicating their intention for this action
-         * 2) if `staker` is a contract, then `signature` must will be checked according to EIP-1271
-         */
-        if (Address.isContract(staker)) {
-            require(IERC1271(staker).isValidSignature(digestHash, signature) == ERC1271_MAGICVALUE,
-                "DelegationManager.delegateToBySignature: ERC1271 signature verification failed");
-        } else {
-            require(ECDSA.recover(digestHash, signature) == staker,
-                "DelegationManager.delegateToBySignature: sig not from staker");
-        }
+        // actually check that the signature is valid
+        EIP1271SignatureUtils.checkSignature_EIP1271(staker, stakerDigestHash, stakerSignatureAndExpiry.signature);
 
-        _delegate(staker, operator);
+        // go through the internal delegation flow, checking the `approverSignatureAndExpiry` if applicable
+        _delegate(staker, operator, approverSignatureAndExpiry);
     }
 
     /**
      * @notice Undelegates `staker` from the operator who they are delegated to.
-     * @notice Callable only by the StrategyManager
+     * @notice Callable only by the StrategyManager.
      * @dev Should only ever be called in the event that the `staker` has no active deposits in EigenLayer.
+     * @dev Reverts if the `staker` is also an operator, since operators are not allowed to undelegate from themselves.
+     * @dev Does nothing (but should not revert) if the staker is already undelegated.
      */
     function undelegate(address staker) external onlyStrategyManager {
         require(!isOperator(staker), "DelegationManager.undelegate: operators cannot undelegate from themselves");
-        delegatedTo[staker] = address(0);
+        address _delegatedTo = delegatedTo[staker];
+        // only make storage changes + emit an event if the staker is actively delegated, otherwise do nothing
+        if (_delegatedTo != address(0)) {
+            emit StakerUndelegated(staker, delegatedTo[staker]);
```

Comment:
_delegatedTo

Reply:
more clearly, `operator`

Reply:
good call. changed in https://github.com/Layr-Labs/eigenlayer-contracts/pull/97/commits/3af75132771abfead6143453ec75efc4e52a23cd